src/item_location.cpp

```diff
@@ -762,6 +762,20 @@ bool item_location::parents_can_contain_recursive( item *it ) const
     return false;
 }
 
+int item_location::max_charges_by_parent_recursive( item it ) const
+{
+    if( !has_parent() ) {
+        return item::INFINITE_CHARGES;
+    }
+
+    item_location parent = parent_item();
+    item_pocket *pocket = parent->contents.contained_where( *get_item() );
+
+    return std::min( { it.charges_per_volume( pocket->remaining_volume() ),
+                       it.charges_per_weight( pocket->remaining_weight() ),
+                       parent.max_charges_by_parent_recursive( it ) } );
```

Comment:
I'm not sure if this logic is correct. Suppose we have a 1.1 L container containing another 1 L container which contains 1 charge of 0.5L liquid. If I call this method on the liquid, the first recursion calculates the maximum charge to be 1 (0.5 L remaining), but the second recursion calculates the maximum charge to be 0 (0.1L remaining in the 1.1L container), so the final result is the 1L container can hold no more charges, while it should be 1.

Reply:
Yes, you're right, this definitely needs to end the recursion if the pocket we're checking is rigid.